Addind CMake based C++ project example in documentation

fraviofii opened the pull request:
## Summary

The objective of this pull request is to provide an application example of a C++ project running outside the NuttX source tree.

The motivation for this documentation is to provide a starting for developers implement C++ based applications using a previously defined NuttX configuration without the need to use the NuttX build structure.

v01d commented on Documentation/guides/cpp_cmake.rst:
```diff
@@ -0,0 +1,274 @@
+
+* hellocpp/CMakeLists.txt
+
+.. code-block::
```
```suggestion
.. code-block:: cmake
```

v01d commented on Documentation/guides/cpp_cmake.rst:
```diff
@@ -0,0 +1,274 @@
+
+* hellocpp/cmake/stm32f4discovery.cmake
+
+.. code-block::
```
```suggestion
.. code-block:: cmake
```

v01d commented on Documentation/guides/cpp_cmake.rst:
```diff
@@ -0,0 +1,274 @@
+
+* hellocpp/src/main.cpp
+
+.. code-block::
```
```suggestion
.. code-block:: c++
```

v01d commented on Documentation/guides/cpp_cmake.rst:
```diff
@@ -0,0 +1,274 @@
+
+* hellocpp/src/HelloWorld.h
+
+.. code-block::
```
```suggestion
.. code-block:: c++
```

v01d reviewed (changes requested):
Some minor changes. Also, can you try using `caption` to set the filename of each piece of code as shown here: https://www.sphinx-doc.org/en/1.4.9/markup/code.html#caption-and-name ? Maybe it looks better.

v01d commented:
Thanks for adding this. Eventually it will be best to organize the guides section into subsections so it does not become a mess, but we can do that later on.

gustavonihei commented on Documentation/guides/cpp_cmake.rst:
```diff
@@ -0,0 +1,287 @@
+support C++ based projects.
+
+Using the 'build as a library' procedure of NuttX, it is possible to build NuttX
+applications using C++ language and also the cmake  build
+tool.
```
```suggestion
applications using C++ language and also the cmake build tool.
```

fraviofii commented:
I have also added an introduction paragraph indicating the motivation for this document.

Are these changes enough?

gustavonihei commented on Documentation/guides/cpp_cmake.rst:
```diff
@@ -0,0 +1,287 @@
+
+    extern "C"
+    {
+            int phigw_main(void)
```
```suggestion
            int hellocpp_main(void)
```
Just to be consistent with the aforementioned Application entry point

v01d commented:
> I have also added an introduction paragraph indicating the motivation for this document.
> 
> Are these changes enough?

Can you try to use the `:caption:` I mentioned?

gustavonihei commented on Documentation/guides/cpp_cmake.rst:
```diff
@@ -0,0 +1,287 @@
+    {
+            int phigw_main(void)
+            {
+                    Log::print("PHI: Exemplo em CMake");
```
```suggestion
```
This log line may be omitted from this example for the sake of simplicity.

gustavonihei commented on Documentation/guides/cpp_cmake.rst:
```diff
@@ -0,0 +1,287 @@
+
+    #include <stdio.h>
+    #include "HelloWorld.h"
+    #include "Log.h"
```
```suggestion
```
Not necessary if the Log class won't be used.

gustavonihei commented on Documentation/guides/cpp_cmake.rst:
```diff
@@ -0,0 +1,287 @@
+    #include "HelloWorld.h"
+    #include "Log.h"
+    #include <nuttx/config.h>
+    #include <iostream>
```
```suggestion
```
<iostream> not needed in this `main.cpp` file.

gustavonihei commented on Documentation/guides/cpp_cmake.rst:
```diff
@@ -0,0 +1,287 @@
+
+.. code-block:: c++
+
+    #include <stdio.h>
```
```suggestion
    #include <cstdio>
```
nit: Since this is a C++ example, how about sticking with the functions from the standard library?

gustavonihei commented on Documentation/guides/cpp_cmake.rst:
```diff
@@ -0,0 +1,283 @@
+
+.. code-block:: c++
+
+    #include <stdio.h>
```
```suggestion
```
This is also not required in `main.cpp`

gustavonihei commented on Documentation/guides/cpp_cmake.rst:
```diff
@@ -0,0 +1,283 @@
+
+    cmake_minimum_required(VERSION 3.2...3.15)
+
+    project(phigw
+            VERSION 1.0
+            DESCRIPTION "Hello world Phi-GW C++ Nuttx"
+    )
```
```suggestion
    project(HelloCpp
            VERSION 1.0
            DESCRIPTION "Hello world C++ Nuttx"
    )
```

v01d commented:
Please squash all commits once you are done adding changes.

gustavonihei commented on Documentation/guides/cpp_cmake.rst:
```diff
@@ -0,0 +1,300 @@
+
+    CHelloWorld::CHelloWorld() {
+            mSecret = 42;
+            printf("Constructor: mSecret=%d\n",mSecret);
+    }
+
+    CHelloWorld::~CHelloWorld() {
+
+    }
+
+    bool CHelloWorld::HelloWorld(void) {
+            printf("HelloWorld: mSecret=%d\n",mSecret);
+
+            std::string sentence = "Hello";
+            printf("TEST=%s\n",sentence.c_str());
+
+            if (mSecret == 42) {
+                    printf("CHelloWorld: HelloWorld: Hello, world!");
+                    return true;
+            }
+            else {
+                    printf("CHelloWorld: HelloWorld: CONSTRUCTION FAILED!\n");
+                    return false;
```
```suggestion
            std::printf("Constructor: mSecret=%d\n",mSecret);
    }

    CHelloWorld::~CHelloWorld() {

    }

    bool CHelloWorld::HelloWorld(void) {
            std::printf("HelloWorld: mSecret=%d\n",mSecret);

            std::string sentence = "Hello";
            std::printf("TEST=%s\n",sentence.c_str());

            if (mSecret == 42) {
                    std::printf("CHelloWorld: HelloWorld: Hello, world!");
                    return true;
            }
            else {
                    std::printf("CHelloWorld: HelloWorld: CONSTRUCTION FAILED!\n");
                    return false;
```
Sorry being to picky on this, but it may fail to compile due to the missing namespace.

fraviofii replied:
Won't it be a problem with NuttX's printf?

Some functions, like printf, seems to be better used not from the language/compiler standard library, but from the platform. That's why I left the printf and stdio.h (and the testcxx example too ...).

In standard C++ project, in general, printf is not used. But I believe in NuttX it is better to use.

gustavonihei replied:
If the C++ library is correctly configured, it shouldn't be a problem.
`std::prinf` is the C++ STL version, which is basically a wrapper for the C language version.

fraviofii commented:
> 
> 
> > I have also added an introduction paragraph indicating the motivation for this document.
> > Are these changes enough?
> 
> Can you try to use the `:caption:` I mentioned?

I added, but the code block content is gone.

xiaoxiang781216 commented:
@fraviofii please squash your patch into one:
```
git rebase --interactive HEAD~18...HEAD
``
and select the squash from terminal.

dagar commented:
Can I gauge if there's any interest in native cmake support for NuttX at this time? I think it would greatly improve integration with 3rd party projects, cross platform development support, and developmer productivity in general (much faster out of tree builds, etc).

acassis commented:
> Can I gauge if there's any interest in native cmake support for NuttX at this time? I think it would greatly improve integration with 3rd party projects, cross platform development support, and developmer productivity in general (much faster out of tree builds, etc).

acassis commented:
> Can I gauge if there's any interest in native cmake support for NuttX at this time? I think it would greatly improve integration with 3rd party projects, cross platform development support, and developmer productivity in general (much faster out of tree builds, etc).

Hi @dagar,will this CMake offer work as an option to our original Makefiles instead of replacing it? The current build system works fine and CMake could be a new tool of preference of some, but it will be a new tool for many developers. Makefile is a consolidated tool.

dagar commented:
> Hi @dagar,will this CMake offer work as an option to our original Makefiles instead of replacing it?

If there was enough interest in the NuttX community I'd propose doing it in parallel in order to fully support everything on all platforms. Then once at feature parity we'd either fully accept it or not. In my experience trying to support two things that do roughly the same thing is a good way to frustrate people and end up with multiple mediocre options.
 
> The current build system works fine and CMake could be a tool of preference of some, but it will be a new tool for many developers.

It's not that kind of tool where it can be left to individual preference, it either works accurately describing how to build the project or it doesn't. The part that can be developer preference is it can generate Makefiles, or Ninja build, or an IDE project. The support would roughly map 1-1 (Make.defs -> CMakeLists.txt per folder), but will ultimately be a bit different.

> Some projects like OpenOCD tried to move from Makefiles to CMake many years ago and it never happened. Makefile is a consolidated tool.

I don't know the history there, but there are plenty of other projects that made the transition (LLVM, Zephyr, PX4, etc). It may be a new tool for some, but it's also the closest thing to a standard build system in the C or C++ world and it solves a ton of headaches with cross platform development.

xiaoxiang781216 commented:
Yes, it's a big change: we have more than 200+ boards, 20000+ source files and 10+ external libraries. It's better to create a new issue to summarize the propose and let the community to dissuss and vote.

fraviofii commented:
> 
> 
> @fraviofii please squash your patch into one:
> 
> ```
> git rebase --interactive HEAD~18...HEAD
> ``
> and select the squash from terminal.
> ```

@xiaoxiang781216 , it is done.
